docs/world_generation.md

```diff
@@ -0,0 +1,75 @@
+
+## Terrain material
+The terrain material depends entirely on the following four factors:
+* Elevation above the ground plane
+* Estimated distance below the terrain surface
+* Temperature
+* Rainfall
+
+Note that temperature and rainfall are noise functions set up for the purpose of allowing varying terrain materials. How these are used is described in detail in `terraingen.rs`, so the details are omitted here.
+
+## The road
+Currently, the only megastructure in Hypermine is a single infinite straight road. This megastructure works by using the state machine `worldgen::NodeStateRoad`, which is much like `worldgen::NodeStateKind`, but instead of defining the ground plane, it defines the plane that divides the road into its "east" and "west" sides (with this terminology assuming that the road runs "north" and "south"). Based on this state, the `worldgen::ChunkParams::generate_road` function will be called for chunks right above the ground plane, generating the road's surface and carving out any terrain that is in the way. The `worldgen::ChunkParams::generate_road_support` function will be called for chunks below the ground plane, generating the wooden truss if the road is above the terrain, acting as a bridge.
+
+## Trees
+For decoration, tiny two-block trees are scattered throughout the terrain, with their density depending on the amount of rainfall. Each tree is generated by placing a wood block next to a dirt or grass block, followed by placing a "leaves" block on that wood block, away from the dirt or grass block. While this algorithm is unaware of gravity, it often generates trees upright because the wood blocks are placed on relatively flat ground. See `worldgen::ChunkParams::generate_trees` for more details.
+
+Note that generating larger trees requires a more complicated algorithm that has not yet been planned out or implemented.
+
+## Implementation details
+Hypermine generates worlds in three main stages:
+* Generate a `NodeState` for each node
+* Generate a `ChunkParams` for each chunk, based on the `NodeState` of all 8 nodes adjacent to the chunk's origin
+* Using the information in `ChunkParams`, asynchronously generate the voxel data for each chunk
+
+The reasoning behind these three phases can be seen most clearly in the noise generation algorithm. The coarse noise function created at the beginning of the algorithm is stored in the `NodeState`. Since the noise needs to be interpolated between adjacent `NodeState`s, the `ChunkParams` needs all 8 adjancent nodes to be constructed. The actual interpolation and the inclusion of additional noise is done in the final, asynchronous phase of the algorithm, which depends entirely on `ChunkParams`.
```

Comment:
Maybe mention why we bother making this async?

Reply:
Alternatively, we could avoid talking about how/when the computation takes place entirely in this section, and focus on the theoretical model.